components/Init/src/virtio_vsock.c

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright 2022, UNSW (ABN 57 195 873 179)
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+#include <stdint.h>
+#include <string.h>
+#include <autoconf.h>
+
+#include <utils/util.h>
+
+#include <camkes.h>
+#include <camkes/dataport.h>
+
+#include <sel4vm/guest_irq_controller.h>
+#include <sel4vm/boot.h>
+
+#include <sel4vmmplatsupport/drivers/virtio_vsock.h>
+
+#include <virtqueue.h>
+#include <camkes/virtqueue.h>
+
+#include <virtio/virtio_vsock.h>
+
+#include "vm.h"
+#include "virtio_vsock.h"
+#include "virtio_irq.h"
+
+static virtio_vsock_t *virtio_vsock;
+
+typedef struct virtio_vsock_cookie {
+    vm_t *vm;
+} virtio_vsock_cookie_t;
+
+/* Configured in camkes */
+#define NUM_PORTS ARRAY_SIZE(socket_layout)
+
+/*
+ * Represents the virtqueues used in a given link between
+ * two VMs.
+ */
+typedef struct conn {
+    int cid;
+    virtqueue_driver_t *send_queue;
+    virtqueue_device_t *recv_queue;
+} conn_t;
+static conn_t connections[NUM_PORTS];
+
```

Comment:
General comment: I'd like to see doxygen style comments on *every* function that say what the function is for, what its arguments are, and any pre/post conditions.